engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/HashCodeAnyNode.java

```diff
@@ -79,14 +80,32 @@ long hashCodeForFloat(float f) {
   }
 
   @Specialization
+  @TruffleBoundary
   long hashCodeForDouble(double d) {
-    return d % 1.0 == 0.0 ? (int) d : Double.hashCode(d);
+    if (d % 1.0 != 0.0) {
+      return Double.hashCode(d);
+    } else {
+      long longValue = Double.valueOf(d).longValue();
+      if (BigIntegerOps.fitsInLong(d)) {
+        return Long.hashCode(longValue);
+      } else {
+        try {
+          return BigDecimal.valueOf(d).toBigIntegerExact().hashCode();
+        } catch (ArithmeticException e) {
+          throw new IllegalStateException(e);
+        }
+      }
+    }
   }
 
   @Specialization
   @TruffleBoundary
   long hashCodeForBigInteger(EnsoBigInteger bigInteger) {
-    return bigInteger.getValue().hashCode();
+    if (BigIntegerOps.fitsInLong(bigInteger.getValue())) {
+      return Long.hashCode(bigInteger.longValue());
+    } else {
+      return bigInteger.getValue().hashCode();
+    }
```

Comment:
Technically this branch should be unreachable, because the constructor of `EBI` has `assert (value.bitLength() > 63);`.

So I think you should be safe to just go for
```suggestion
    return bigInteger.getValue().hashCode();
```

if we ever change this invariant we'd need to adapt, but IMO that would be fair.

Reply:
Right. This seems to work. Let's just integrate your suggestion and deal with possible errors once they arrive in the future.